src/System.ComponentModel.TypeConverter/src/System/ComponentModel/BindingList.cs

```diff
@@ -0,0 +1,735 @@
+        private bool _allowEdit = true;
+        private bool _allowRemove = true;
+        private bool _userSetAllowNew = false;
+
+        #region Constructors
+
+        /// <summary>
+        ///     Default constructor.
+        /// </summary>
+        public BindingList() : base()
+        {
+            Initialize();
+        }
+
+        /// <summary>
+        ///     Constructor that allows substitution of the inner list with a custom list.
+        /// </summary>
+        public BindingList(IList<T> list) : base(list)
+        {
+            Initialize();
+        }
+
+        private void Initialize()
+        {
+            // Set the default value of AllowNew based on whether type T has a default constructor
+            _allowNew = ItemTypeHasDefaultConstructor;
+
+            // Check for INotifyPropertyChanged
+            if (typeof(INotifyPropertyChanged).IsAssignableFrom(typeof(T)))
+            {
+                // Supports INotifyPropertyChanged
+                _raiseItemChangedEvents = true;
+
+                // Loop thru the items already in the collection and hook their change notification.
+                foreach (T item in this.Items)
+                {
+                    HookPropertyChanged(item);
+                }
+            }
+        }
+
+        private bool ItemTypeHasDefaultConstructor
+        {
+            get
+            {
+                Type itemType = typeof(T);
+
+                if (itemType.IsPrimitive)
+                {
+                    return true;
+                }
+
+                if (itemType.GetConstructor(BindingFlags.Public | BindingFlags.Instance | BindingFlags.CreateInstance, null, new Type[0], null) != null)
+                {
+                    return true;
+                }
+
+                return false;
+            }
+        }
+
+        #endregion
+
+        #region AddingNew event
+
+        /// <summary>
+        ///     Event that allows a custom item to be provided as the new item added to the list by AddNew().
+        /// </summary>
+        public event AddingNewEventHandler AddingNew
+        {
+            add
+            {
+                bool allowNewWasTrue = AllowNew;
+                _onAddingNew += value;
+                if (allowNewWasTrue != AllowNew)
+                {
+                    FireListChanged(ListChangedType.Reset, -1);
+                }
+            }
+            remove
+            {
+                bool allowNewWasTrue = AllowNew;
+                _onAddingNew -= value;
+                if (allowNewWasTrue != AllowNew)
+                {
+                    FireListChanged(ListChangedType.Reset, -1);
+                }
+            }
+        }
+
+        /// <summary>
+        ///     Raises the AddingNew event.
+        /// </summary>
+        protected virtual void OnAddingNew(AddingNewEventArgs e)
+        {
+            if (_onAddingNew != null)
+            {
+                _onAddingNew(this, e);
+            }
```

Comment:
Nit: just another example where C#6 simplifies things (in fact, the IDE has a refactoring for this):
```C#
_onAddingNew?.Invoke(this, e);
```